src/lib/constants.ts

```diff
@@ -15,6 +15,11 @@ export const SOCKET_IO_PING_CODE = '2';
 export const DEFAULT_RECONNECT_LIMIT = 20;
 export const DEFAULT_RECONNECT_INTERVAL_MS = 5000;
 export const UNPARSABLE_JSON_OBJECT = {};
+export const DEFAULT_HEARTBEAT = {
+  kind: 'ping',
+  timeout: 10000,
+  interval: 25000,
```

Comment:
Doesn't `interval > timeout` mean this will not do what a heartbeat is supposed to do?

If you only send a "ping" every 25 seconds, you'll only get a "pong" from the server every 25 seconds if there's no other messages going back or forth... and with a timeout of 10 seconds, this will disconnect before the next "pong" comes in even if the server is very prompt in responding to "ping"s, right?

Reply:
This is an excellent catch. The `timeout` before closing should be allowed to be shorter than the ping `interval`, but the timer should only start when the ping is sent. I think something like this should work:

```ts
import { DEFAULT_HEARTBEAT } from "./constants";
import { HeartbeatOptions } from "./types";

export function heartbeat(ws: WebSocket, options?: HeartbeatOptions): void {
  const {
    interval = DEFAULT_HEARTBEAT.interval,
    timeout = DEFAULT_HEARTBEAT.timeout,
    kind = DEFAULT_HEARTBEAT.kind,
  } = options || {};

  let messageAccepted = false;
  let closeTimeout;

  const resetCloseTimeout = () => {
    if (closeTimeout) {
      clearTimeout(closeTimeout);
      closeTimeout = null;
    }
  };

  //My implementation in the other PR would require a slightly different implementation, but the idea would be the same.
  ws.addEventListener("message", () => {
    resetCloseTimeout();
  });

  const pingTimer = setInterval(() => {
    try {
      ws.send(kind);
      resetCloseTimeout();

      closeTimeout = setTimeout(() => {
        ws.close();
      }, timeout);
    } catch (error) {
      // do nothing
    }
  }, interval);

  ws.addEventListener("close", () => {
    clearInterval(pingTimer);
    resetCloseTimeout();
  });
}
```

Reply:
I'd agree it should be allowed, but it's probably the intention of the default settings to have the interval at 10s and the timeout at 25s, right? :sweat_smile:
Because that would mean the heartbeat would tolerate 1 failed heartbeat, without closing the connection, but if two fail in a row, then that's a confirmed dead connection ("the heart has stopped beating")